### PR TITLE
📱 Add a 'Mobile behavior' selector on /admin/layout #1237

### PR DESCRIPTION
**Pull request by ithiame**
Add a "Mobile behavior" selector on /admin/layout #1237

### CODE REVIEW

File: src/routes/+layout.svelte [hunk 1]
```diff
@@ -18,6 +19,15 @@
...
 	<title>{data.websiteTitle}</title>
+	{#if data.roleId !== CUSTOMER_ROLE_ID && data.viewportFor === 'employee'}
+		<meta name="viewport" content="width=device-width" />
+	{:else if data.roleId === CUSTOMER_ROLE_ID && data.viewportFor === 'visitors'}
+		<meta name="viewport" content="width=device-width" />
+	{:else if data.viewportFor === 'everyone'}
+		<meta name="viewport" content="width=device-width" />
+	{:else}
+		<meta name="viewport" content="width={data.viewportContentWidth}" />
+	{/if}
```
Comment (coyotte508) [inline src/routes/+layout.svelte:30]: Let's move this to the backend, the backend should  send `viewportWidth` depending on the role.

Here we should only have : 
```suggestion
	<meta name="viewport" content="width={data.viewportWidth}" />
```

This means that you will need to send both `viewportFor` and `viewportContentWidth` in the admin's `+page.server.ts`

File: src/routes/+layout.server.ts [hunk 2]
```diff
@@ -1,6 +1,24 @@
...
+	) {
+		viewportWidth = 'width=device-width';
+	} else if (
+		event.locals.user?.roleId === CUSTOMER_ROLE_ID &&
+		runtimeConfig.viewportFor === 'visitors'
+	) {
+		viewportWidth = 'width=device-width';
+	} else {
+		viewportWidth = runtimeConfig.viewportContentWidth;
+	}
```
Comment (coyotte508) [inline src/routes/+layout.server.ts:20]: use a `switch`:

```ts
const viewPortWidth = () => ({
  switch (runtimeConfig.viewportFor) {
    case ...: return ... ? ... : ...;
    ...
  }
})()
```

`switch` has better type-checking.

Also makes the logic clearer, I think some cases are wrong with the current logic (for example `'visitors` should include when `roleId` is `undefined`)

**Comment by Tirodem** [comment]
@ithiame :
❌  one too many "width=" :
image
❌ I cannot save when selector value is "everyone" (no error on console, "OK" on confirm nothing happens) ; it only saves when I click "cancel" on confirm, in opposition of the other values 🤔

File: src/routes/+layout.server.ts [hunk 3]
```diff
@@ -1,23 +1,24 @@
...
+		switch (runtimeConfig.viewportFor) {
+			case 'everyone':
+				return 'width=device-width';
+			case 'employee':
+				return event.locals.user?.roleId !== CUSTOMER_ROLE_ID
+					? 'width=device-width'
+					: runtimeConfig.viewportContentWidth;
+			case 'visitors':
+				return event.locals.user?.roleId !== POS_ROLE_ID &&
+					event.locals.user?.roleId !== SUPER_ADMIN_ROLE_ID
```
Comment (coyotte508) [inline src/routes/+layout.server.ts:15]: Some employees can have another role id.

You should check instead:

- roleId === undefined || roleId = CUSTOMER_ROLE_ID

File: src/routes/+layout.server.ts [hunk 4]
```diff
@@ -1,23 +1,24 @@
...
-		viewportWidth = 'width=device-width';
-	} else {
-		viewportWidth = runtimeConfig.viewportContentWidth;
-	}
+	const viewportWidth = (() => {
+		switch (runtimeConfig.viewportFor) {
+			case 'everyone':
+				return 'width=device-width';
+			case 'employee':
+				return event.locals.user?.roleId !== CUSTOMER_ROLE_ID
```
Comment (coyotte508) [inline src/routes/+layout.server.ts:10]: Guests (not logged in) have `roleId = undefined`, with this code, they have the same view as employees

File: src/routes/+layout.server.ts [hunk 5]
```diff
@@ -1,6 +1,26 @@
...
+import { CUSTOMER_ROLE_ID } from '$lib/types/User';
 
 export async function load(event) {
+	const viewportWidth = (() => {
+		switch (runtimeConfig.viewportFor) {
+			case 'everyone':
+				return 'width=device-width';
+			case 'employee':
+				return event.locals.user?.roleId !== CUSTOMER_ROLE_ID ||
+					event.locals.user?.roleId !== undefined
```
Comment (coyotte508) [inline src/routes/+layout.server.ts:11]: This condition will always be `true`

File: src/routes/+layout.server.ts [hunk 6]
```diff
@@ -1,6 +1,26 @@
...
+					? 'width=device-width'
+					: runtimeConfig.viewportContentWidth.toString();
+			case 'visitors':
+				return event.locals.user?.roleId === undefined ||
+					event.locals.user?.roleId === CUSTOMER_ROLE_ID
+					? 'width=device-width'
+					: runtimeConfig.viewportContentWidth.toString();
```
Comment (coyotte508) [inline src/routes/+layout.server.ts:18]: It's the opposite no?

**Comment by Tirodem** [comment]
- "Everyone" selection is now ok ✅ 
- Selecting "Employee" gives `width=device-width` to employees but also to non-employees ❌ (cf https://github.com/B2Bitcoin/beBOP/pull/1238#discussion_r1585374423 )
- About https://github.com/B2Bitcoin/beBOP/pull/1238#discussion_r1585374869 , functional behavior seems good 🤔 (when selecting "Visitors", visitors have `width=device-width` and employees have the standard view

For personal notes & future documentation, about viewport width :
- header / footer go mobile when width < 640px
- product page goes mobile when width < 768px
- /checkout page goes mobile when width < 768px
- /order page goes mobile when width < 768px
- /cart doesn't seem to have significant change

**Comment by Tirodem** [comment]
+ @ithiame as you removed the 2nd "width=" for width=device-width, we don't have "width" anymore for viewport size, so it doesn't work :
image
I'm pushing a quick modification to allow @CypherJack to do some tests today and tomorrow.

**Comment by ithiame** [comment]
My Bad 
let me change it quickly

File: src/routes/+layout.server.ts (hunk 6 above)
Comment (ithiame) [reply src/routes/+layout.server.ts]: it's correct i think 

??

**Review by coyotte508 — APPROVED** [review]

**Pull request merged**